Timeline : Draw the frame number next to the playhead

themissingcow opened the pull request:
Straw person implementation for consideration.

![image](https://user-images.githubusercontent.com/896779/59437742-0529cd00-8dea-11e9-8623-49251908319a.png)

User facing changes :

 - The Timeline will now display the current frame number next to the
   playhead.

 - The Timeline will preview the frame number under the cursor when
   hovering over the timeline area to make frame selection easier.

Due to the way `paintEvent` is implemented in `Slider` I couldn't find a sensible way to scope this to a `Timeline` specific implementation. As such, this the introduces the concept of mouse interaction snapping to `Slider` and consequently `NumericSlider`.  It its implemented such that its effect is scoped to mouse actions, rather than all position/value sets.

TODO: What level of test coverage do we want for something like this?

andrewkaufman commented:
Off topic: @johnhaddon this is a Draft PR and both Travis and Azure ran... any idea how its different to Cortex?

lucienfostier commented on python/GafferUI/Timeline.py:
```diff
@@ -259,5 +259,34 @@ def __init__( self, value=None, min=0, max=1, hardMin=None, hardMax=None, values
+
+		textX = sliderX + 6
+		if textX + frameTextSize.width() + 6 >= size.x :
+			textX = sliderX - frameTextSize.width() - 6
```
can you explain the +/-6 logic, if any :)
perhaps you could use constant variable to avoid the magic number syndrome?

themissingcow replied:
Good call, does https://github.com/GafferHQ/gaffer/pull/3208/commits/e13e52708f582e82200e43b78ba0962b29e37aae help?

mattigruener commented:
Hey Tom,

I like that this is bringing `AnimationGadget` and `Timeline` closer together in terms of UX :) For the sake of consistency, would it make sense to

- use the same color for the gray frame indicator? ( currently it's `Color3f( 120 / 255.0 )` in the AnimationGadget as opposed to using opacity and white for the `Timeline` )
- adjust the style even further so that we render that little yellow flag with dark text on the `Timeline`? Not sure if there's even enough space...
- improve the `AnimationGadget` side of things to also include the number for the gray frame indicator?

These are just suggestions, of course :) Thanks for all your UI work recently, I like these changes :)

Matti.

medubelko commented:
Hi Tom, just following up on your fastidious labelling: which existing docs need to be updated for this? Were you thinking of the static screenshots?

themissingcow commented:
> Hi Tom, just following up on your fastidious labelling: which existing docs need to be updated for this? Were you thinking of the static screenshots?

Hey, I'd just tagged it more as it 'changed the UX'. I wasn't sure if there were any places where the Timeline was talked about so was just tagging it so we didn't miss it if you felt it appropriate to make any changes. If you're happy, then nothing to do.

themissingcow commented:
> * use the same color for the gray frame indicator? ( currently it's `Color3f( 120 / 255.0 )` in the AnimationGadget as opposed to using opacity and white for the `Timeline` )

Done (though ultimately, I think the opacity is a nicer presentation, but one to push back to when we improve the `AnimationGadget` I think?).

> * adjust the style even further so that we render that little yellow flag with dark text on the `Timeline`? Not sure if there's even enough space...

I gave it a bash, there isn't enough space for it to work visually sadly :(

> * improve the `AnimationGadget` side of things to also include the number for the gray frame indicator?

Yeah that's a good idea. Would you mind if kept this scoped to just the `Timeline` though? It was a 'had quick half hour on a Friday' thing so was relatively low priority. I think we might need to give the `AnimationGadget` a once over when we need to support other curve interpolation types, so could look at it then?

> These are just suggestions, of course :) Thanks for all your UI work recently, I like these changes :)

They're good suggestions!  Thanks :) glad its to taste!

themissingcow commented:
I've rebased this to latest, and fixed an issue where we'd draw the 'hover' appearance over the highlighted one in a drag (which I'd not noticed before due to the opacity, so double win Matti!).

mattigruener commented:
> Done (though ultimately, I think the opacity is a nicer presentation, but one to push back to when we improve the AnimationGadget I think?).

> I think we might need to give the AnimationGadget a once over when we need to support other curve interpolation types, so could look at it then?

Totally. I don't think there's a need to do this right now. It's an good improvement already :) Thanks!

johnhaddon commented on python/GafferUI/Slider.py:
```diff
@@ -206,6 +214,64 @@ def getPositionIncrement( self ) :
+	## Sets whether mouse interactions with the slider should snap to
+	# specific position increments. This does not apply to programmatic
+	# changes via setPosition(s) or keyboard increments.
+	def setSnaps( self, snaps ) :
```
`setSnaps()` seems like an awkward phrasing, but I think we do seem to generally be avoiding the verbosity of something like `setSnappingEnabled()`. In similarly awkward situations we have things like `ViewportGadget::setDragTracking()` and `ViewportGadget::setPlanarMovement()`. I think that implies this could maybe be `setSnapping()`? Happy to hear your opinion, but I would like to aim for consistency and guessability where we can.

themissingcow replied:
Removed in rebase/rework

johnhaddon commented on python/GafferUI/Slider.py:
```diff
@@ -206,6 +214,64 @@ def getPositionIncrement( self ) :
+
+		return self.__snaps
+
+	def setSnapIncrement( self, snapIncrement ) :
```
Alternatively, we could have `snapIncrement is None` be the way of signalling that snapping is off?

themissingcow replied:
Switched to an increment of `None` meaning no snapping in rework.

johnhaddon commented on python/GafferUI/Slider.py:
```diff
@@ -206,6 +214,64 @@ def getPositionIncrement( self ) :
+
+		return self._snapOverridable
+
+	def setDrawHoverPosition( self, draw ) :
```
setHoverPositionVisible?

themissingcow replied:
Good idea

themissingcow commented on python/GafferUI/NumericSlider.py:
```diff
@@ -128,12 +129,29 @@ def valueChangedSignal( self ) :
+
+		return self.__min + ( position * ( self.__max - self.__min ) )
+
+	def _positionForValue( self, value ) :
```
Used by `Timeline` as `_drawPosition` is given a position, and has no means to convert it back to a value.

themissingcow replied:
Symmetry of `_valueForPosition`, I guess from past experience consuming APIs, I'm not a massive fan of asymmetrical access.

themissingcow commented on python/GafferUI/NumericSlider.py:
```diff
@@ -128,12 +129,29 @@ def valueChangedSignal( self ) :
+
+		return round( value / self._snapIncrement ) * self._snapIncrement
+
+	def _valueForPosition( self, position ) :
```
Used by `Timeline` as `_drawPosition` is given a position, and has no means to convert it back to a value.

themissingcow commented on python/GafferUI/NumericSlider.py:
```diff
@@ -128,12 +129,29 @@ def valueChangedSignal( self ) :
+
+	def _snapValue( self, value ) :
+
+		return round( value / self._snapIncrement ) * self._snapIncrement
```
(Not fully resurfaced all of the reasoning in my head yet but) I guess it seemed like it made sense (to my silly brain at least) to say that when you configure a `Slider`, snap values are in positions, as thats all that they have. When you configure a `NumericSlider`, then they are all in numeric space. Otherwise, afaict, you'd have to make all the `valueToPosition` methods public too (as opposed to the requested even-more-private) so people could set meaningful values. 2p

themissingcow commented on python/GafferUI/NumericSlider.py:
```diff
@@ -128,12 +129,29 @@ def valueChangedSignal( self ) :
+		value = self._valueForPosition( position )
+		return self._positionForValue( self._snapValue( value ) )
+
+	def _snapValue( self, value ) :
```
Just in the sense that `_snapPosition` is protected such that derived classes such as `NumericSlider` can re-implement snapping to their value space, `_snapValue` would allow classes derived from `NumericSlider` to affect their own value snapping if desired.  If we want to be really conservative with what we expose until we have an actual concrete use case, it can totally be private.

themissingcow commented on python/GafferUI/Slider.py:
```diff
@@ -73,6 +74,12 @@ def __init__( self, position=None, positions=None, **kw ) :
 
+		self._snaps = snaps
+		self._snapIncrement = snapIncrement
+		self._snapOverridable = snapOverrideable
```
Def can fix these, I think it was an oversight when adding the accessors for them, and forgot to add the extra `_`

themissingcow commented on python/GafferUI/Slider.py:
```diff
@@ -206,6 +214,64 @@ def getPositionIncrement( self ) :
+
+		return self._drawHoverPosition
+
+	def _shouldSnap( self, reason, event = None ) :
```
Called in the `QWidget`'s `paintEvent` via `owner._shouldSnap` so had to be protected.

themissingcow commented:
@johnhaddon I've rebased and refactored this on to your #4081 `Slider` refactor. I've incorporated some of the suggestions you made last time around in terms of naming, etc. The one thing I have left-as-is for now is the `protected`/`private` debate. Derived classes such as `Timeline` need `_positionToValue` in `_drawPosition`. Can move the others to private if required, but many are also used by the `QtWidget` derived helper class, so it is also about providing `friend` access. We have a mix of `protected` and mangled access in that widget already, so this way, at least it was using one convention. Let me know if you want me to make them all `private` aside from the ones currently required in `Timeline`.

themissingcow commented:
This is pr-hold untill #4081 is merged.

johnhaddon commented:
> @johnhaddon I've rebased and refactored this on to your #4081 Slider refactor. I've incorporated some of the suggestions you made last time around in terms of naming, etc. The one thing I have left-as-is for now is the protected/private debate. Derived classes such as Timeline need _positionToValue in _drawPosition.

I've refactored again to double down on making stuff private, so that the only remaining protected members are the two drawing functions which can be overridden by derived classes. I also refactored the clamping and hover code so that the addition of snapping was a trivial addition with all the logic in a single place. I also rejigged the drawing method so that derived classes have more information and no need to access state. The user-visible result should be the same as before, but I think this leaves things in a more maintainable state.
